### PR TITLE
Add "Go To Definition"

### PR DESCRIPTION
**Pull request by mint-thompson**
Completes task [CIMPL-340](https://standardhealthrecord.atlassian.net/browse/CIMPL-340).

When in a workspace, enable "Go To Definition". The extension assumes that the first entity it finds with the target name is the desired entity. This should hopefully, usually be correct. Because the extension now makes use of typescript, tsconfig is added, and the package file specifies an entry point into the code.

The latest version of the FSH grammar at time of writing is used.

The launch configuration present in the repo should be suitable for testing locally, but let me know if you have questions about how to test. You may need to add a second command-line argument to the launch configuration specifying a folder to open, since the "Go To Definition" feature only works when there is an active workspace.

### CODE REVIEW

File: package.json
Comment (jafeltra) [inline package.json:58]: In order for the `npm run compile` script to run, I had to save `typescript` as a dev dependency. Were you able to get this to work without adding it? I also only added it to `devDependencies` because that's where it was in the CIMPL extension.

**Review by jafeltra — COMMENTED** [review]
This looks really awesome! I tried it out by using the "Go to Definition" feature within the mCODE repo and tried out a combinations of the various definition types we have, and most seem to be working as expected. I just had a few pretty minor questions inline, and the couple case I found that didn't seem to work as expected are below:

- Go to Definition from an obeys rule to an Invariant seems to not be working for me. Does it work for you?
- Similar, the Go to Definition on a simple RuleSet doesn't seem to navigate to the RuleSet definition. This did work for a parameterized RuleSet, and also it might have been something I did wrong.

File: package.json
Comment (mint-thompson) [reply package.json:53]: Probably! I'll add it in.

File: package.json
Comment (mint-thompson) [reply package.json:58]: Hmm, that might be because I have typescript installed globally, and that installation got used. But if it's needed, it should be added, so I'll add it.

Comment (jafeltra) [reply package.json:60]: Hmm now that you mention that, I realized I also have it installed globally, so I wonder why it didn't use that also.

**Review by jafeltra — APPROVED** [review]
This looks awesome to me! Those last two cases that I tried earlier now work, and I can't think of any other definitions to go to! This is a really useful feature!

**Comment by jafeltra** [comment]
Oh one other thought I just had. Do you think we should include anything in the README file about the new "Go to Definition" feature? I believe the README is what shows up in the Extension Marketplace, so it might be helpful to show off our new feature there! What do you think?

**Review by cmoesel — CHANGES_REQUESTED** [review]
This is very cool.  I think people will find it quite helpful!  I did run into a few issues, however.

**Issues Related to Development**

1. When I tried to run the extension at first, I got an error launching the extension:
image
It only worked after I ran `npm run compile` (which, btw, I had to ctl-c out of because it is in watch mode). I expect you can reproduce this by deleting your `out` folder and then attempting to run the extension.  I don't know if it is possible, but ideally, the launch configuration would automatically build it as part of the launch -- otherwise I might make changes and forget to compile, then think I'm debugging the changed source (when I'm actually not). If that's not possible, can you document that the source needs to be built before you run and/or document that best practice is to keep a watching build in another console as you develop?

2. In order to have some consistency w/ our other projects, can you rename the `compile` script to `build:watch` and add a `build` script that just does the build once? And can you document these scripts in the README?

3. Now that this is officially a typescript project, can we add the same `lint` and `prettier` rules as our other projects -- along w/ their corresponding scripts (`lint`, `lint:fix`, `prettier`, `prettier:fix`, `check`)? (And again, add the documentation to the README)

4. It's not clear how you got the ANTLR4-generated files.  Did you just copy them from SUSHI?  If so can you document the process?  If not, can we add the infrastructure we need to get repeatable generation of the ANTLR4 files (like in SUSHI's `antlr` directory)?

**Issues Related to Use**

5. As @jafeltra suggested, I think it would be good to document this new (awesome) feature in the README.

6. I think that maybe the change you made to address the issues that Julia noticed introduced bugs in other areas.  For example, it doesn't work with references (e.g., `Reference(CancerPatient)` does not properly find the definition for `CancerPatent`). I'm guessing it is because you are searching the whole phrase, including `Reference(` ... `)`.

7. Now or Later: I've noticed it does properly go to the definition if I right-click on an `id`. It seems it only works with `name`. We can save this for later, but eventually it would be nice to support `id` too.  Supporting `url` would also be wicked cool, but... that gets a lot more complicated.

8. Later: I've also noticed that it allows me to right-click on _anything_ and say `Go to Definition`.  Technically, we know, based on the grammar, where supported "target" phrases are -- so I think we _could_ only show the menu item where it makes sense grammatically.  But we probably shouldn't invest in that now.

9. Later: It looks like it is parsing on every invocation of `Go to Definition` (if I understand it correctly).  It would be cool to get better performance w/ caching, but I expect that's complex since we'd need to track what files have changed and what files have not.  But something to consider...

10. Probably Later: I wonder if we it might simplify things (in some cases) if we just brought in SUSHI as a dependency.  This "Go to Definition" feature would then be as simple as running the SUSHI import and calling the fisher functions on the `FshTank` (since we put the source locators in the class instances).  The downside is additional overhead and the caching solution mentioned in 9 above would be a lot more complicated since we'd need to implement it inside of SUSHI.

So... I'd prefer we try to address 1 - 6 as part of this PR, but 7-10 can wait unless you think any of them are easy enough to slip in.

Can't wait for users to get their hands on this!

**Comment by mint-thompson** [comment]
@cmoesel I've addressed the first six points from above. But, there are some things that I'd like to add, either as part of this PR, or as part of a follow-up task.
- It would be nice to have an animated gif of "Go to Definition" working, like we have for snippets. Do you have a recommended tool for making this?
- Adding in eslint and prettier made me think that there should also be automated tests for this extension. That way, scenarios like misbehaving on `Reference(MyProfile)` will be more plainly visible. I've done some investigation into how to add tests for extensions, so the only question is whether that's in scope for this task. I would say it is, for the same reason that adding eslint and prettier were in scope for this task. There is a question of whether to use mocha (which is what is used in the vscode extension documentation, and appears to be what is recommended) or jest (which can definitely be used with a little extra set-up work, and is what we're already using on other FSH projects). Do you have an opinion about which test library to use?

**Comment by cmoesel** [comment]
>     * ld be nice to have an animated gif of "Go to Definition" working, like we have for snippets. Do you have a recommended tool for making this?

I'm on Mac, so any tools I recommend probably wouldn't apply.  I actually used CleanShotX, which is a paid app.  If you can't find a good (free) solution for Windows, I can make an animated gif for you!

I'm not going to argue against tests -- but I also assume they are hard/complicated.  Is that true?  If so, how much time do you think it will take to do this?  As for the test framework -- I think most JS developers are pretty familiar with `mocha`, so if that is what VS Code documentation recommends and/or uses as examples, I'm fine with it.  I guess I don't really have a strong preference there -- as you can make a case either way.

**Review by jafeltra — COMMENTED** [review]
The changes look good! I had a few small comments inline. Also, I'm not sure if I tested this one and it worked before, but right now, if there's an alias used in a code, like `* dosage.route = SCT#26643006 "Oral use"`, when I try to Go to Definition of `SCT`, it doesn't bring me to the Alias. I think it should since it's an alias we defined.

File: .vscode/launch.json
Comment (mint-thompson) [reply .vscode/launch.json:15]: ugh, I thought I had remembered to _not_ add that to the commit. You are correct that this should not be in there.

File: .vscode/tasks.json
Comment (mint-thompson) [reply .vscode/tasks.json:2]: This is the version of the VSCode tasks file being used. The documentation for tasks doesn't directly refer to this, but from the page on [converting from 0.1.0 to 2.0.0](https://code.visualstudio.com/Docs/editor/tasks#_convert-from-010-to-200), it seems like "2.0.0" is the current version string to use.

File: README.md
Comment (mint-thompson) [reply README.md:50]: Nope! This should be at the H2 level.

**Review by jafeltra — COMMENTED** [review]
The tests look really awesome! It's cool to have automated tests and having them working with GH Actions! I had one really small question on them below.

Also, I went back to your first commit to see if the Alias in a coding Go to Definition was working, and it looks like it was 😕 But if it seems like a lot of work to get it working now, it's probably okay to hold off on that one for now.

**Review by jafeltra — APPROVED** [review]
This looks so awesome. The last case that I had noticed is fixed now (and works in the new unit test you added - yay!). I looked through Chris's comments and I thought 1-6 were addressed, but I can also let him confirm that as well. I didn't notice anything else, so I'm re-approving! Awesome feature!

**Review by cmoesel — APPROVED** [review]
Very cool.  I love it!  Thank you, @mint-thompson!

**Comment by cmoesel** [comment]
I'm going to force merge this.  The checks have not run because they are not yet in the master branch.  I think after this PR gets merged, the checks in future PRs should run just fine.

**Pull request merged**
